src/Serialise/Cborg/Tutorial.hs

```diff
@@ -16,33 +16,239 @@ module Serialise.Cborg.Tutorial
+> encodeAnimal (WalkingAnimal name speed) =
+>     encodeListLen 3 <> encodeTag 1 <> encode name <> encode speed
+
+Here we see that each encoding begins with a /length/, describing how many
+values belonging to our @Animal@ will follow. We then encode a /tag/, which
+identifies which constructor. We then encode the fields.
 
 -}
 
 {- $decoding
 
-Lorem ipsum...
+Decoding CBOR representations to Haskell values is done in the 'Decoder'
+'Monad'. We can write a 'Decoder' for the @Animal@ type defined above as
+follows,
+
+> decodeAnimal :: Decoder s Animal
+> decodeAnimal = do
+>     len <- decodeListLen
+>     tag <- decodeTag
+>     case (len, tag) of
+>       (3, 0) -> HoppingAnimal <$> decode <*> decode
+>       (3, 1) -> WalkingAnimal <$> decode <*> decode
+>       _      -> fail "invalid Animal encoding"
+
+-}
+
+{- $migrations
+
+One eventuality that data serialisation schemes need to account for is the need
+for changes in the data's structure. There are two types of compatibility
+which we might want to strive for in our serialisers,
+
+ * Backward compatibility, such that newer versions of the serialiser can read
+   older versions of an encoding
+
+ * Forward compatibility, such that older versions of the serialiser can read
+   (or at least tolerate) newer versions of an encoding
+
+Below we will look at a few of the types of
+changes which we may need to make and describe how these can be handled in a
+backwards-compatible manner with @cborg@.
+
+*** Adding a constructor
+
+Say we want to add a new constructor to our @Animal@ type, @SwimmingAnimal@,
+
+> data Animal = HoppingAnimal { animalName :: String, hoppingHeight :: Int }
+>             | WalkingAnimal { animalName :: String, walkingSpeed :: Int }
+>             | SwimmingAnimal { numberOfFins :: Int }
+>             deriving (Generic)
+
+We can account for this in our hand-rolled serialiser by simply adding a new tag
+to our encoder and decoder,
+
+> encodeAnimal :: Animal -> Encoding
+> {- ... other cases are unchanged -}
+> encodeAnimal (SwimmingAnimal numberOfFins) =
+>     encodeListLen 2 <> encodeTag 2 <> encode numberOfFins
+>
+> decodeAnimal :: Decoder s Animal
+> decodeAnimal = do
+>     len <- decodeListLen
+>     tag <- decodeTag
+>     case (len, tag) of
+>       {- ... other cases are unchanged -}
+>       (2, 2) -> SwimmingAnimal <$> decode
+>       _      -> fail "invalid Animal encoding"
+
+*** Adding\/removing\/modifying fields
+
+Say then we want to add a new field to our @WalkingAnimal@ constructor,
+
+> data Animal = HoppingAnimal { animalName :: String, hoppingHeight :: Int }
+>             | WalkingAnimal { animalName :: String, walkingSpeed :: Int, numberOfFeet :: Int }
+>             | SwimmingAnimal { numberOfFins :: Int }
+>             deriving (Generic)
+
+We can account for this by representing @WalkingAnimal@ with a new encoding with
+a new tag,
+
+> encodeAnimal :: Animal -> Encoding
+> {- ... HoppingAnimal, SwimmingAnimal cases are unchanged -}
```

Comment:
Here again the comment isn't shown, and should just include the code it's describing.